cijoe/scripts/bench_plotter.py

```diff
@@ -0,0 +1,463 @@
+
+    collection = []
+
+    search = Path(search)
+    for path in search.rglob(f"{BDEVPERF_OUTPUT_PREFIX}*.txt"):
+        with path.open() as ofile:
+            match = re.match(OUTPUT_REGEX, str(path))
+
+            sample = {
+                "ctx": {},
+                "iops": 0,
+                "bwps": 0,
+                "lat": 0,
+                "stddev": 0,
+            }
+
+            for ctx in ["bs", "iodepth", "label"]:
+                if ctx == "iodepth":
+                    sample["ctx"][ctx] = int(match.group(ctx))
+                else:
+                    sample["ctx"][ctx] = match.group(ctx)
+            sample["ctx"]["name"] = sample["ctx"]["label"]
+
+            for line in ofile.readlines():
+                metrics = re.match(LINE_REGEX, line)
+                if metrics:
+                    for metric in ["iops", "bwps"]:
+                        sample[metric] = float(metrics.group(metric))
+
+            m = hashlib.md5()
+            m.update(
+                str.encode(
+                    "".join(
+                        [
+                            f"{key}={val}"
+                            for key, val in sorted(sample["ctx"].items())
+                            if key not in ["timestamp", "stem", "iodepth"]
+                        ]
+                    )
+                )
+            )
+
+            collection.append((str(m.hexdigest()), sample))
+
+    artifacts = search / "artifacts"
+
+    with (artifacts / BDEVPERF_OUTPUT_NORMALIZED_FILENAME).open("w") as jfd:
+        json.dump(collection, jfd, **JSON_DUMP)
+
+    return 0
+
+
+def extract(args, cijoe, step):
+    """
+    Traverse 'args.output' for 'fio-output_*.txt' files, and create:
+
+    * A 'fio-output_*.json' per 'fio-output_*.txt'
+    * A single "compound" 'fio-output-compound.json' containing all the discovered
+      JSON-documents with the file.stem as key.
+    """
+
+    search = step.get("with", {}).get("path", args.output)
+    if not search:
+        return errno.EINVAL
+
+    compound = {}
+
+    search = Path(search)
+    for path in search.rglob("fio-output_*.txt"):
+        jsondict = dict_from_fio_output_file(path)
+        with path.with_suffix(".json").open("w") as jfd:
+            json.dump(jsondict, jfd, **JSON_DUMP)
+
+        compound[path.stem] = jsondict
+
+    artifacts = search / "artifacts"
+
+    with (artifacts / FIO_COMPOUND_FILENAME).open("w") as jfd:
+        json.dump(compound, jfd, **JSON_DUMP)
+        log.info(f"wrote: {jfd.name}")
+
+    return 0
+
+
+def normalize(args, cijoe, step):
+    """Transform the "compound" fio output into series of 'normalized' data"""
+
+    def contextualize(fio_output: dict, stem) -> dict:
+        """Helper function for the transformation of output"""
+
+        assert len(fio_output["jobs"]) == 1
+
+        job = fio_output["jobs"][0]
+        options = job["job options"]
+
+        # should io-engine-args be added here? They probably should...
```

Comment:
I don't think this comment adds anything

Reply:
Right